### DISCUSSION TITLE
libmdbx on gramine

### DISCUSSION THREAD

**Discussion by tiagorvmartins** [gramineproject/gramine#2068] [Q&A] [answered]
Hey Gramine team, would appreciate some assistance,
Been trying to run an application that depends on libmdbx, already enabled experimental flock but still having issues.

**The error seems to be:**
mdbx_env_open: function not implemented

Which I believe is trying to do some kind of file locking unsupported on Gramine?

Some supporting info here:
[mdbx_env_open](https://erthink.github.io/libmdbx/group__c__opening.html#gabb7dd3b10dd31639ba252df545e11768) 

Below, a few debug log lines when running on gramine:

```
[INFO] [11-21|20:16:48.095] Set global gas cap                       cap=50000000
[INFO] [11-21|20:16:48.098] [Downloader] Running with                ipv6-enabled=true ipv4-enabled=true download.rate=16mb upload.rate=4mb
[INFO] [11-21|20:16:48.099] Opening Database                         label=data path=/tmp/data
PATH:  /tmp/data
FLAGS:  41943040
(libos_stat.c:185:__do_statfs) [P1:T1:app] debug: statfs: 20000000 10000000 10000000
(libos_stat.c:185:__do_statfs) [P1:T1:app] debug: statfs: 20000000 10000000 10000000
mdbx_env_open: function not implemented
(libos_context.c:279:prepare_sigframe) [P1:T1:app] debug: Created sigframe for sig: 23 at 0xc00000f390 (handler: 0x4947c0, restorer: 0xa5e97ef2e0)
[EROR] [11-21|20:16:48.136] App startup                           err="mdbx_env_open: function not implemented, label: data, trace: [kv_mdbx.go:361 node.go:367 node.go:370 backend.go:270 node.go:131 main.go:73 make_app.go:46 command.go:276 app.go:333 app.go:307 main.go:46 proc.go:267 asm_amd64.s:1650]"
mdbx_env_open: function not implemented, label: data, trace: [kv_mdbx.go:361 node.go:367 node.go:370 backend.go:270 node.go:131 main.go:73 make_app.go:46 command.go:276 app.go:333 app.go:307 main.go:46 proc.go:267 asm_amd64.s:1650]
(libos_exit.c:205:libos_syscall_exit_group) [P1:T1:app] debug: ---- exit_group (returning 1)
(libos_async.c:109:install_async_event) [P1:T2:app] debug: Installed async event at 1732220208137588
(libos_async.c:109:install_async_event) [P1:T8:app] debug: Installed async event at 1732220208137588
(libos_async.c:109:install_async_event) [P1:T7:app] debug: Installed async event at 1732220208137607
```

Any clues what can be done?
Thank you!

**Comment by kailun-qin** [comment]
@tiagorvmartins: Can you pls provide the "trace" Gramine logs? Thanks!

**Reply by tiagorvmartins** [reply]
Hey @kailun-qin thank you for your reply, attached the trace logs,
Thank you.

[not_supported_trace.log](https://github.com/user-attachments/files/17856316/not_supported_trace.log)

**Reply by kailun-qin** [reply]
Looking at the log, `(libos_parser.c:1701:buf_write_all) [P1:T1:cdk-erigon] trace: ---- madvise(0x4b3f205000, 0xfb000, MADV_DONTFORK) = -38` looks suspicious to me. This is currently unsupported in Gramine: https://github.com/gramineproject/gramine/blob/e1405529111807ba2f3fad8fcb911b1e5d8f38d1/libos/src/sys/libos_mmap.c#L505-L513

From a quick look at libmdbx, this might happen during `mdbx_env_open()` -> `mdbx_setup_dxb()`-> `mdbx_mmap()` -> `madvise(MADV_DONTFORK)`: https://github.com/erthink/libmdbx/blob/bb8f43181783686879219846d64379a04c1430e3/src/osal.c#L1530-L1533). As a workaround, can you pls try disabling `MDBX_ENABLE_MADVISE` build option: https://github.com/erthink/libmdbx/blob/bb8f43181783686879219846d64379a04c1430e3/src/options.h#L83-L88 or undef `MADV_DONTFORK` in libmdbx?

**Reply by tiagorvmartins** [reply]
Hey @kailun-qin thank you for the fast reply,

I was able to change the build and disabled MDBX_ENABLE_MADVISE, not geetting madvise(...) = -38 anymore but getting instead now the following error, which might be another indicator/hint:
`warning: PI futexes are not yet supported!`

Is there anything else that I could disable/enable or easily swap here?
Also attaching more trace logs here of this last run:
[not_supported_trace_2.log](https://github.com/user-attachments/files/17868262/not_supported_trace_2.log)

Thank you!

**Reply by kailun-qin** [reply]
Yes, the log shows `futex(0xb577a8394c, FUTEX_PRIVATE|OP 7, 0, 0, 0xc, 0) = -38` however `OP 7`/`FUTEX_UNLOCK_PI` is currently unsupported:
https://github.com/gramineproject/gramine/blob/e1405529111807ba2f3fad8fcb911b1e5d8f38d1/libos/src/sys/libos_futex.c#L797-L803

I think you'll need to look into the source code of libmdbx or the pthread library to see how to fall back to other solutions.

**Reply by tiagorvmartins** [reply]
Hey @kailun-qin I happen to use another build flags and eventually was able to use another locking system,
But now I am facing another issue, which was unexpected,
Question: isn't pwritev supported in gramine?
Looked at the docs and I believe it is, but wanted to double check.

Getting on trace logs:
` warning: Unsupported system call pwritev`

Thank you!

**Reply by kailun-qin** [reply]
> Question: isn't pwritev supported in gramine?

No, it's currently unsupported: https://github.com/gramineproject/gramine/blob/5789620a6d04208e5803321953e09eca2c24dd7b/libos/src/arch/x86_64/libos_table.c#L313

**Reply by tiagorvmartins** [reply]
Okay I see now that explains it, but docs are outdated then I guess?
I see this in the docs:

☑ pwritev() [9a](https://gramine.readthedocs.io/en/latest/devel/features.html#file-system-operations)

Thank you!

**Reply by kailun-qin** [reply, marked as answer]
> Okay I see now that explains it, but docs are outdated then I guess?

Yeah, I noticed that discrepancy in the doc -- it's not correct.

**Reply by tiagorvmartins** [reply]
Okay thank you, I managed to work around it using a wrapper around pwritev leveraging pwrite64.
For now seems to have solved the problem, thank you again @kailun-qin !
